src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala

```diff
@@ -75,49 +130,129 @@ object RockSlicer {
-      (s(medianIndex), s.take(medianIndex), s.drop(medianIndex + 1))
+    if (remainingBlocks.nonEmpty) {
+      // Merged blocks still contain some processor joints
+      mergeBlocks(remainingBlocks ++ processorBlocks.tail, completedBlocks ++ mergedBlocks, origin)
+    } else if (processorBlocks.tail.isEmpty) {
+      // All blocks are free of processor joints - check for duplicates then return
+      val mergedBlocksDuplicates = completedBlocks ++ mergedBlocks
+      mergedBlocksDuplicates.foldLeft(Seq[Block]()) { (unique, current) =>
+        if (!unique.exists(Block.compareBlocks(current, _)))
+          current +: unique
+        else
+          unique
+      }
+    } else {
+      // Proceed to next processor block
+      mergeBlocks(processorBlocks.tail, completedBlocks ++ mergedBlocks, origin)
     }
+  }
+
+  /**
+    * Compares two input blocks and determines whether they share a processor face
+    * @param block1 First input block
+    * @param block2 Second input block
+    * @return List of processor faces that are shared by the two blocks. Will be 
+    *         empty if they share no faces
+    */
+  def compareProcessorBlocks(block1: Block, block2: Block): Seq[Face] = {
+    val processorFaces1 = block1.faces.filter { case face => face.processorJoint }
+    val processorFaces2 = block2.faces.filter { case face => face.processorJoint }
+    val faceMatches = 
+      processorFaces1 map { case face1 =>
+        processorFaces2 map { case face2 =>
+          if ((math.abs(face1.a + face2.a) < NumericUtils.EPSILON) &&
+              (math.abs(face1.b + face2.b) < NumericUtils.EPSILON) &&
+              (math.abs(face1.c + face2.c) < NumericUtils.EPSILON) &&
+              (math.abs(face1.d + face2.d) < NumericUtils.EPSILON)) {
+            Seq[Face](face1, face2)
+          } else Seq[Face]()
+        }
+      }
+    faceMatches.filter{ case faces => faces.nonEmpty }.flatten.flatten.distinct
+  }
+
+  /**
+    * Compares two input faces and determines whether faces are shared. Shared
+    * faces will have equal and opposite distances from local origin as well as
+    * normal vectors in opposite directions
+    * @param face1 First input face
+    * @param face2 Second input face
+    * @return True if faces are shared, false otherwise
+    */
+  def compareSharedFaces(face1: Face, face2: Face): Boolean = {
+    (math.abs(face1.a + face2.a) < NumericUtils.EPSILON) &&
+    (math.abs(face1.b + face2.b) < NumericUtils.EPSILON) &&
+    (math.abs(face1.c + face2.c) < NumericUtils.EPSILON) &&
+    (math.abs(face1.d + face2.d) < NumericUtils.EPSILON)
+  }
 
-    val queue = new mutable.PriorityQueue[Seq[Int]]()(Ordering.by(_.length))
-    queue.enqueue(indices)
-    loadBalanceIndicesImpl(queue)
+  // Function that writes JSON string to file for single node - taken from 
+  // http://stackoverflow.com/questions/4604237/how-to-write-to-a-file-in-scala
+  private def printToFile(f: java.io.File)(op: java.io.PrintWriter => Unit) {
```

Comment:
Hmm, this code is pretty weird to me. There are more straightforward ways to accomplish this, but I actually think we should just stick to Spark's utilities anyways.